src/WebJobs.Script/Host/HostIdValidator.cs

```diff
@@ -0,0 +1,198 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Storage.Blobs;
+using Microsoft.Azure.WebJobs.Hosting;
+using Microsoft.Azure.WebJobs.Logging;
+using Microsoft.Azure.WebJobs.Script.Properties;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using IApplicationLifetime = Microsoft.AspNetCore.Hosting.IApplicationLifetime;
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    /// <summary>
+    /// Used to perform Host ID validation checks, ensuring that when hosts are sharing
+    /// a storage account, their computed IDs don't collide.
+    /// </summary>
+    /// <remarks>
+    /// <see cref="ScriptHostIdProvider"/> computes a Host ID and truncates it if needed to
+    /// ensure it's under length limits. For two different Function Apps, this can result in
+    /// both apps resolving to the same Host ID. This can cause problems if those apps share
+    /// a storage account. This class helps detect/prevent such cases.
+    /// </remarks>
+    public class HostIdValidator
+    {
+        public const string BlobPathFormat = "ids/usage/{0}";
+        private const LogLevel DefaultLevel = LogLevel.Warning;
+
+        private readonly IEnvironment _environment;
+        private readonly IAzureStorageProvider _storageProvider;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IApplicationLifetime _applicationLifetime;
+        private readonly HostNameProvider _hostNameProvider;
+        private readonly ILogger _logger;
+
+        private readonly object _syncLock = new object();
+        private bool _validationScheduled;
+
+        public HostIdValidator(IEnvironment environment, IAzureStorageProvider storageProvider, IApplicationLifetime applicationLifetime,
+            HostNameProvider hostNameProvider, ILoggerFactory loggerFactory, IServiceProvider serviceProvider)
```

Comment:
Inject an `ILogger<HostIdValidator>` here instead of the factory (trying to get away from those custom categories we started with).